Sources/Workspace/Workspace.swift

```diff
@@ -1707,30 +1714,60 @@ extension Workspace {
                 targetName: artifact.targetName
             ]
 
-            if let existingArtifact = existingArtifact, case .remote(_, let existingChecksum) = existingArtifact.source {
-                // If we already have an artifact with the same checksum, we don't need to download it again.
-                if artifact.checksum == existingChecksum {
-                    continue
-                }
+            if let existingArtifact = existingArtifact {
+                switch existingArtifact.source {
+                case .remote(url: _, checksum: let existingChecksum):
+                    // If we already have an artifact with the same checksum, we don't need to download it again.
+                    if artifact.checksum == existingChecksum {
+                        continue
+                    }
 
-                // If the checksum is different but the package wasn't updated, this is a security risk.
-                if !addedOrUpdatedPackages.contains(artifact.packageRef) {
-                    diagnostics.emit(.artifactChecksumChanged(targetName: artifact.targetName))
-                    continue
-                }
+                    // If the checksum is different but the package wasn't updated, this is a security risk.
+                    if !addedOrUpdatedPackages.contains(artifact.packageRef) {
+                        diagnostics.emit(.artifactChecksumChanged(targetName: artifact.targetName))
+                        continue
+                    }
 
-                artifactsToRemove.append(existingArtifact)
+                    artifactsToRemove.append(existingArtifact)
+                case .localArchived:
+                    artifactsToRemove.append(existingArtifact)
+                case .local:
+                    break
+                }
             }
 
             artifactsToDownload.append(artifact)
         }
 
+        for artifact in manifestArtifacts.localArchived {
+            let existingArtifact = self.state.artifacts[
+                packageIdentity: artifact.packageRef.identity,
+                targetName: artifact.targetName
+            ]
+
+            if let existingArtifact = existingArtifact {
+                switch existingArtifact.source {
+                case .remote:
+                    artifactsToRemove.append(existingArtifact)
+                case .localArchived:
+                    // TODO: It is better to compare the file signatures of the archives to optimize the
+                    // performance by reusing the existing artifact instead of re-extracting the old one
+                    // each time.
+                    artifactsToRemove.append(existingArtifact)
```

Comment:
Here we may want to reuse the already existing managed artifact by additionally persisting its file signature and comparing it with the potential artifact (just as it is currently implemented for the remote artifacts, but with the automatically generated signature).

Reply:
It points out the outdated line now.